### PR TITLE
Upgrade Gradle and Android SDK versions.

### PR DESCRIPTION
**Pull request by msand**
## Motivation
**Address Path Traversal Vulnerability**
Improve build speeds by restricting which dependencies leak their APIs to other modules.
Improve performance on up to date Android phones.
Decrease memory use.

https://developer.android.com/studio/releases/gradle-plugin.html#3-0-0
https://developer.android.com/studio/build/gradle-plugin-3-0-0-migration.html#new_configurations

To maintain your application along with each Android release, you should increase the value of this attribute to match the latest API level, then thoroughly test your application on the corresponding platform version.
https://developer.android.com/guide/topics/manifest/uses-sdk-element.html

https://medium.com/google-developers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd

https://developer.android.com/training/basics/supporting-devices/platforms.html#sdk-versions

Align with FBSDK:
[Facebook SDK for Android Changelog 4.x](https://developers.facebook.com/docs/android/change-log-4x)

> Changelog and release notes for the Facebook SDK for Android.
> 4.30.0 - January 24, 2018
> Facebook SDK
> Added
> 
>     Protection against Path Traversal Vulnerability: https://support.google.com/faqs/answer/7496913
>     Support for variant-aware dependency management in Gradle 4.1 and Android Studio 3.0

[Path Traversal Vulnerability](https://support.google.com/faqs/answer/7496913)
What’s happening

Beginning January 16th, 2018, Google Play will block publishing of any new apps or updates which contain the Path traversal Vulnerability. Your published APK version will remain unaffected, however any updates to the app will be blocked unless you address this vulnerability.

[Android SDK versions](https://developer.android.com/reference/android/os/Build.VERSION_CODES.html?utm_campaign=adp_series_sdkversion_010616&utm_source=medium&utm_medium=blog)
https://developer.android.com/sdk/api_diff/24/changes.html
https://developer.android.com/sdk/api_diff/25/changes.html
https://developer.android.com/sdk/api_diff/26/changes.html
https://developer.android.com/sdk/api_diff/27/changes.html

https://developer.android.com/about/versions/oreo/android-8.0.html
Various media api improvements
Picture-in-Picture mode
Partial Java 8 support
https://developer.android.com/studio/write/java8-support.html

https://developer.android.com/about/versions/oreo/android-8.1.html
> Memory optimizations. Improved memory usage across the platform to ensure that apps can run efficiently on devices with 1GB or less RAM.

> Flexible targeting options. New hardware feature constants to let you target the distribution of your apps to normal or low-RAM devices through Google Play.

> Google Play.While all apps will be available on devices running Android Oreo (Go edition), Google Play will give visibility to apps specifically optimized by developers to provide a great experience for billions of people with the building for billions guidelines.

## Test Plan

Run CI

## Related PRs
[Update docs](https://github.com/facebook/react-native-website/pull/156)

## Release Notes
[ANDROID] [ENHANCEMENT] [Framework] - Upgrade to latest stable and hopefully less vulnerable versions.

### CODE REVIEW

**Comment by hramos** [comment]
Thanks for opening the PR. We want to make sure the Android SDK version we use here matches the one we use at Facebook. I'll see what has been done internally and I will get back to this.

**Comment by nitaliano** [comment]
@hramos any news on this?

**Comment by hramos** [comment]
Not yet. Current priority is getting all of our tests back to green. Until then, we cannot make this sort of change to the codebase.

**Comment by nitaliano** [comment]
@hramos are you referring to internal fb tests? If you're talking about external facing tests is there any place where we can all break them down and hand them out to people to help get them to start passing? I think staying up to date with the native tools are very important so I would be happy to help in anyway that I can.

**Comment by hramos** [comment]
I was referring to external tests. They run on Circle, you can find the definitions at `.circleci/config.yml`. These are now green as of last week.

We do have a large suite of tests that run internally, that for the most part concern the use of React Native in our own product's surfaces. I am currently focusing on catching test breakages internally at Facebook before they land and sync to GitHub. There are several reasons this may happen even with our internal tests in place: we use our own test infrastructure that is quite different from Circle; we may be using newer buck builds that have yet to reach open source; we have a monorepo that depends on the latest Metro packager which may also not yet exist in npm; and so on. 

While we use buck to build the project internally, we do not do so from within the same open source React Native repo. I am working on resolving that blind spot, in order to catch buck failures that have broken our Circle tests in the past.

Which brings me to this PR: as I am adding buck back to our internal tests, I am running into issues due to the use of different Android SDK versions between internal/open source. Once I solve this, I should come back to this PR and suggest which exact versions we should be using in order to match internal.
 
Stay tuned.

File: RNTester/android/app/build.gradle [hunk 1]
```diff
@@ -135,9 +134,9 @@ android {
 }
 
 dependencies {
-    compile fileTree(dir: 'libs', include: ['*.jar'])
-    compile 'com.android.support:appcompat-v7:23.0.1'
+    api fileTree(dir: 'libs', include: ['*.jar'])
+    api 'com.android.support:appcompat-v7:27.0.2'
```
Comment (hramos) [inline RNTester/android/app/build.gradle]: (y) We're using 27.0.2 internally.

Comment (msand) [reply RNTester/android/app/build.gradle]: Good to know.

**Comment by MarkOSullivan94** [comment]
@hramos looks like Circle CI is needing to upgrade the emulator and download the Android SDK 27 for it to pass test_android

> Your emulator is out of date, please update by launching Android Studio:
>  - Start Android Studio
>  - Select menu "Tools > Android > SDK Manager"
>  - Click "SDK Tools" tab
>  - Check "Android Emulator" checkbox
>  - Click "OK"


> Error: could not find version  of the Android SDK.
> Specifically, the directory /opt/android/sdk/platforms/android- does not exist.
> You probably need to specify the right version using the SDK Manager from within Android Studio.

File: ReactAndroid/build.gradle [hunk 2]
```diff
@@ -231,12 +231,11 @@ task packageReactNdkLibsForBuck(dependsOn: packageReactNdkLibs, type: Copy) {
 }
 
 android {
-    compileSdkVersion 23
-    buildToolsVersion "23.0.1"
```
Comment (charpeni) [inline ReactAndroid/build.gradle]: Actually, we rely on `buildToolsVersion` in the CI. That's why `test_android` is failing.

https://github.com/facebook/react-native/blob/b63015d34ef51585c452142f666d4f36ceaa1b56/scripts/validate-android-test-env.sh#L55

**Comment by charpeni** [comment]
@MarkOSullivan94 Thanks for the heads up. 

The error is located here: https://github.com/facebook/react-native/pull/17747/#discussion_r185937805. 

Also, it looks like we may have to clear the CircleCI cache, because new Android deps are not installed if `$ANDROID_HOME/installed-dependencies` is there.

File: ReactAndroid/build.gradle (hunk 2 above)
Comment (msand) [reply ReactAndroid/build.gradle]: The new gradle version will choose the build tools version itself, but specifying it here won't (afaik) hurt either.

File: ReactAndroid/build.gradle [hunk 3]
```diff
@@ -292,7 +291,7 @@ android {
 dependencies {
     compile fileTree(dir: 'src/main/third-party/java/infer-annotations/', include: ['*.jar'])
     compile 'javax.inject:javax.inject:1'
-    compile 'com.android.support:appcompat-v7:23.0.1'
+    api 'com.android.support:appcompat-v7:27.0.2'
     compile 'com.facebook.fbui.textlayoutbuilder:textlayoutbuilder:1.0.0'
```
Comment (gengjiawen) [inline ReactAndroid/build.gradle]: replace compile with implementation

File: ReactAndroid/src/androidTest/AndroidManifest.xml [hunk 4]
```diff
@@ -3,7 +3,7 @@
           package="com.facebook.react.tests.gradle"
           android:versionCode="1"
           android:versionName="1.0" >
-  <uses-sdk android:targetSdkVersion="7" />
+  <uses-sdk android:targetSdkVersion="27" />
```
Comment (gengjiawen) [inline ReactAndroid/src/androidTest/AndroidManifest.xml]: This is useless.

**Comment by gengjiawen** [comment]
@msand Any time to resolve the conflict ?

**Comment by MarkOSullivan94** [comment]
I haven't looked into the source code of [create-react-native-app](https://github.com/react-community/create-react-native-app) in too much detail so perhaps someone here would know the answer to my question: How will this change affect create-react-native-app?

I know the `eject` command builds the `android` and `ios` directories for RN projects but how would it know what gradle version / target API version to use?

#18095 would also be affecting by this.

File: ReactAndroid/src/androidTest/AndroidManifest.xml (hunk 4 above)
Comment (tasomaniac) [reply ReactAndroid/src/androidTest/AndroidManifest.xml]: should be completely removed. Also libraries should not even define `targetSdk`

**Comment by MarkOSullivan94** [comment]
I think Circle CI is broken. The checks it's showing are from 1 month ago: https://circleci.com/gh/facebook/react-native/41794?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Current status: 4 failing and 5 successful checks.

**Comment by tasomaniac** [comment]
Don't you need to update react.gradle file too? That all depends on implementation details of the old version of Android tools.

**Comment by msand** [comment]
@gengjiawen @tasomaniac Fixed conflicts and addressed feedback

**Comment by dulmandakh** [comment]
closing this PR because SDK 27, support library 27.x and android gradle plugin 3.1 is landed in master.

**Pull request closed without merging**